### PR TITLE
Spacing Support: Update padding support to allow configurable sides

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
## Description

This builds upon https://github.com/WordPress/gutenberg/pull/30606 which updates the BoxControl to allow configurable sides. 

In this PR, the spacing support is updated to accept this new configuration. Its inclusion into the `hooks/style.js` is refactored a little as well to bring it in line with other block supports and smooth the addition of margin support in the next PR of this chain.

## How has this been tested?
Manually.

#### Test Instructions
1. Checkout this PR
2. Create a post, adding a group block to it. Also adding a background color makes viewing the padding changes easier.
3. Select the group block and adjust its padding via the sidebar controls. Ensure the padding BoxControl still works as expected.
4. Change the group block's padding support configuration to only allow certain sides e.g. `"padding": [ "top", "bottom" ]`
    Screen Shot 2021-04-09 at 5 12 23 pm
5. Reload the editor. Re-select the group block and confirm the padding box control now only allows configuration of the sides you selected.
6. Save the post and confirm the correct padding values applied to the group block on the frontend

#### Test Instruction - Global Styles
1. Ensure you have a block based theme active and open the site editor
2. If there is no group block present add one
3. Update your theme.json configuration adding settings for the `core/group` block context opting into padding support but only for top/bottom sides.
4. Confirm there are no spacing controls under Root in the Global Styles sidebar
5. Switch to the "By Block Type" tab in the Global Styles sidebar
6. Expand the Group block and you should see the padding box control allowing top/bottom values to be set
7. Update your theme.json `styles` to add `core/group` block context, specifying padding values.
8. Reload the editor and ensure the box control under the Group block styles reflect your theme.json values.
9. Modify these padding values and ensure the Reset button works. https://github.com/WordPress/gutenberg/pull/30606#discussion_r609504393

## Screenshots 
| All Sides | Custom Sides |
|---|---|
| PaddingCustomSidesSupport-NoCustomisation | PaddingCustomSidesSupport-CustomSides |

## Types of changes
New feature. Enhancement.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/block-editor/src/hooks/spacing.js
Comment (youknowriad) [inline packages/block-editor/src/hooks/spacing.js:50]: Looks like this could use the `hasPaddingSupport` from the padding.js file? like done with `useIsPaddingDisabled`

File: packages/edit-site/src/components/sidebar/spacing-panel.js
Comment (youknowriad) [inline packages/edit-site/src/components/sidebar/spacing-panel.js:7]: For another time in the future, it looks like the global styles UI could share a lot of code with the hooks (not specific to this PR). cc @nosolosw

**Review by youknowriad — COMMENTED** [review]
This PR is looking very good. We just need to figure the format of the "sides" and land both these two PRs.

File: packages/block-editor/src/hooks/spacing.js
Comment (aaronrobertshaw) [reply packages/block-editor/src/hooks/spacing.js:50]: I had this separate so in the next PR all that was needed was the [addition of `|| support?.margin`](https://github.com/WordPress/gutenberg/pull/30608/files#diff-e819cbce5f99d1503f31d9cdae0f87b00c8e76315f683e98b2cde792f92379d8R51)

I could replace this approach with calls to `hasPaddingSupport` and `hasMarginSupport` if that is better. Whatever you'd prefer is fine by me.

File: packages/edit-site/src/components/sidebar/spacing-panel.js
Comment (aaronrobertshaw) [reply packages/edit-site/src/components/sidebar/spacing-panel.js:7]: On the topic of sharing, there could be some benefit for the block support hooks to be able to access the default theme.json style values e.g. to reset controls correctly. From memory, I don't think these values are currently available within the post editor's state like they are in the site editor. 

So if a RangeControl is reset it might just be to `0` however the managed styles created from the theme.json might be `50`. The result is visually the control doesn't really match what is seen in the editor.

File: packages/block-editor/src/hooks/spacing.js
Comment (youknowriad) [reply packages/block-editor/src/hooks/spacing.js:50]: I don't mind either, I just thought since you shared the hook maybe it makes sense to do the same here; But I don't feel strongly.

Comment (aaronrobertshaw) [reply packages/block-editor/src/hooks/spacing.js:50]: I've updated this to use `hasPaddingSupport`, it will be quicker to grok now.

**Comment by youknowriad** [comment]
This could use a rebase, now that we merged the configurable sides for the component.

**Comment by aaronrobertshaw** [comment]
This has been rebased after the merge of the updated BoxControl. 

The other PRs using this branch as their base have also been rebased accordingly. (https://github.com/WordPress/gutenberg/pull/30608, https://github.com/WordPress/gutenberg/pull/30609)

File: packages/edit-site/src/components/sidebar/spacing-panel.js
Comment (youknowriad) [inline packages/edit-site/src/components/sidebar/spacing-panel.js]: We already  have a `useEditorFeature` hook to be used instead of `useThemeValues` which does the same thing but doesn't take care of the entire chain of theme.json

**Review by youknowriad — COMMENTED** [review]
Left a comment about useEditorFeature that I think is important. Other than that this looks good, I think we should probably add a paragraph about this to  `block-supports.md` documentation.

File: packages/edit-site/src/components/sidebar/spacing-panel.js
Comment (oandregal) [reply packages/edit-site/src/components/sidebar/spacing-panel.js:7]: > For another time in the future, it looks like the global styles UI could share a lot of code with the hooks (not specific to this PR). cc @nosolosw

I'd like to see that very much. I think the current struggle is that the hooks depend on the block's attributes as input ([example](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/hooks/color.js#L215)).

File: packages/edit-site/src/components/sidebar/spacing-panel.js
Comment (aaronrobertshaw) [reply packages/edit-site/src/components/sidebar/spacing-panel.js]: The idea here was to be able to reset the control in the Global Styles sidebar to the actual values set by the theme. 

Using `useEditorFeature` we get restricted to just the features in `__experimentalFeatures` rather than being able to access the styles as well. Testing `useEditorFeature` in place of `useThemeValues` here, results in `undefined`  instead of theme set values, when setting padding values under the default theme.json's "styles" property for the `core/group` context.

```json
	"styles": {
		"core/group": {
			"spacing": {
				"padding": {
					"top": "50px",
					"bottom": "50px"
				}
			}
		}
	}
```

In earlier iterations that evolved into this PR, I used a ref to reset the control back to the initial values at the time the site editor was loaded. It was pointed out that it would be more in line with user expectations to reset to whatever the theme set prior to any adjustments the user made.

If there is a better approach to meet the goal of resetting to the theme values, I'm happy to make it happen.

**Comment by aaronrobertshaw** [comment]
> Other than that this looks good, I think we should probably add a paragraph about this to `block-supports.md` documentation.

I've updated the `block-supports.md` to illustrate that sides can be configured for padding support now.

All that is left I believe is to get a confirmation on what should be changed regarding `useEditorFeature` vs `useThemeValues` in the `spacing-panel.js`

**Comment by aaronrobertshaw** [comment]
This has been rebased after the [changes on trunk to padding.js and spacing-panel.js ](https://github.com/WordPress/gutenberg/commit/f818f1d7562e87ca551777bb4c286e29b1d8f617)fixing custom CSS units.

File: packages/edit-site/src/components/sidebar/spacing-panel.js
Comment (youknowriad) [reply packages/edit-site/src/components/sidebar/spacing-panel.js]: I understand but I think we should be cautious here to not introduce an adhoc solution, this is a problem for all properties, not just padding and right now, resetting means unsetting the property for all the other properties, so I think  we should simplify and just do that for the moment.

Comment (oandregal) [reply packages/edit-site/src/components/sidebar/spacing-panel.js]: From https://github.com/WordPress/gutenberg/pull/30608#discussion_r618268847

I understand that the goal of `useThemeValues( name, property )` is to get the base values (core+theme). I believe is in our best interest to evolve `getStyles` to do this instead. For example, updating it so it can do `getStyle( name, property, 'theme' )` seems a fine choice. The rationale for that is that we want to centralize access to the shape of the data, so consumers don't need to deal with shape changes, migrations, etc. ― instead, there's a centralized place for all of that (global styles provider, getStyle method, etc).

Comment (oandregal) [reply packages/edit-site/src/components/sidebar/spacing-panel.js]: That sort of change to the current API will be a great well-scoped PR that can land independently from these two.

Comment (aaronrobertshaw) [reply packages/edit-site/src/components/sidebar/spacing-panel.js]: Agreed. Thanks as always for the wise counsel @youknowriad and @nosolosw 👍 

I'll update the padding and margin support PRs to simply unset values for the time being. Then next week I'll put together a PR to update `getStyle` to offer the `theme` origin.

Comment (aaronrobertshaw) [reply packages/edit-site/src/components/sidebar/spacing-panel.js]: This has now been updated to unset padding values rather than reset to the theme values. `useThemeValues` has been removed. https://github.com/WordPress/gutenberg/pull/30607/commits/2567acfb8f5519ff4bb6d8322d03bb0203bb1ad3

File: packages/edit-site/src/components/sidebar/spacing-panel.js
Comment (youknowriad) [inline packages/edit-site/src/components/sidebar/spacing-panel.js:103]: Should we  hide the panel entirely if there's no padding control?

**Review by youknowriad — APPROVED** [review]
This looks good to me. @nosolosw any thoughts?

File: packages/edit-site/src/components/sidebar/spacing-panel.js
Comment (aaronrobertshaw) [reply packages/edit-site/src/components/sidebar/spacing-panel.js:103]: Unless I've missed something, we already do.

Within `global-styles-sidebar.js` there is a check using `useHasSpacingPanel` as to whether the `SpacingPanel` should be displayed. This matches the current approach for the typography, color and border panels.

The `showPaddingControl` check here within the component is because we'll soon be [adding a control for margins](https://github.com/WordPress/gutenberg/pull/30608/files?file-filters%5B%5D=.js#diff-bb729c1a11a1203794da39c490560c4c9ac0031906d9981794b9d32d372f1d63R138-R146) as well.

Comment (youknowriad) [reply packages/edit-site/src/components/sidebar/spacing-panel.js:103]: I guess I was just wondering why this component (taken separately) renders an empty panel if padding is not shown (I'm sure this will change with the addition of margins though). This is not that important.

File: packages/block-editor/src/hooks/spacing.js
Comment (oandregal) [inline packages/block-editor/src/hooks/spacing.js]: What would you think of inlining `SpacingPanelControl`? This is: remove that component and use the sub-components directly in here, as we do in the typography hook et all. This has some code duplication that's difficult to reason about (we check for isDisabled and isSpacingEnabled in two places: here and within `SpacingPanelControl`).

File: packages/edit-site/src/components/sidebar/spacing-panel.js
Comment (oandregal) [inline packages/edit-site/src/components/sidebar/spacing-panel.js]: It looks like this function is exactly the same than `useCustomSides` in the hook, so I wonder if we can consolidate. Fine to do it later as well.

File: docs/reference-guides/block-api/block-supports.md
Comment (oandregal) [inline docs/reference-guides/block-api/block-supports.md:506]: Was `padding` shipped in WordPress 5.7?

- If so, we'll need to see how to support the old key as well.
- If not, this will still merit at least a breaking change note in the block-editor package (and bump its version accordingly).

**Comment by oandregal** [comment]
The only "blocker" for me would be to clarify https://github.com/WordPress/gutenberg/pull/30607/files#r621172978

File: packages/edit-site/src/components/sidebar/spacing-panel.js
Comment (aaronrobertshaw) [reply packages/edit-site/src/components/sidebar/spacing-panel.js]: A PR is up adding a `theme` origin to `getStyle`. https://github.com/WordPress/gutenberg/pull/31267

File: packages/edit-site/src/components/sidebar/spacing-panel.js
Comment (aaronrobertshaw) [reply packages/edit-site/src/components/sidebar/spacing-panel.js]: Thanks, that [duplication has been removed](https://github.com/WordPress/gutenberg/pull/30607/commits/5d1b856b72a59558b67a1761e4fda75cad3ae4cb). I've exported `useCustomSides` and used this in `spacing-panel.js`

File: docs/reference-guides/block-api/block-supports.md
Comment (aaronrobertshaw) [reply docs/reference-guides/block-api/block-supports.md:506]: It’s quite possible I’ve gotten lost on this one. To me, when the API was stablized and then ready to be documented, it should have been `spacing.padding` within this doc file all along?

The padding block support flag was made stable in https://github.com/WordPress/gutenberg/pull/25788 changing from `__experimentalPadding` to `spacing.padding`. That PR was part of the Gutenberg v9.2 release and WordPress 5.7 included up to v9.9 of Gutenberg.

Happy to be pointed in the right direction if that is not correct as well as make any required changes.

File: packages/block-editor/src/hooks/spacing.js
Comment (aaronrobertshaw) [reply packages/block-editor/src/hooks/spacing.js]: Definitely an improvement. Done ✅ 

It was a remnant from when I originally set out to later reuse the panel in the global styles sidebar.

File: docs/reference-guides/block-api/block-supports.md
Comment (oandregal) [reply docs/reference-guides/block-api/block-supports.md:506]: Thanks for sharing that. I've checked and this is fine, it's actually fixing a bug we had in the docs.

It looks like we used `spacing.padding` when it was made stable, but the docs were updated incorrectly https://github.com/WordPress/gutenberg/pull/26771 https://github.com/WordPress/gutenberg/pull/26859

**Review by oandregal — APPROVED** [review]
Nice one!

**Comment by oandregal** [comment]
I see there are some conflicts but they're simply [import issues](https://github.com/WordPress/gutenberg/pull/30607/conflicts).

**Comment by oandregal** [comment]
Just discovered the "conflicts" tool from GitHub so I went ahead and fixed those conflicts ― it's nice! This is ready to merge once the tests pass.

**Comment by aaronrobertshaw** [comment]
Thanks for all the help on this 🙇 

Hopefully by proxy this has resolved a lot of the issues for adding margin support in https://github.com/WordPress/gutenberg/pull/30608 and it should be an easier review.

**Pull request merged**
